File: quant1x/data/kline_raw.cpp
#include "kline_raw.h"
#include <quant1x/factors/base.h>
#include <quant1x/config/cache.h>
#include <quant1x/std/filepath.h>

namespace data {

    namespace detail {
        // 拉取数据
        std::vector<level1::SecurityBar> fetch_kline(const std::string &code, u16 start, u16 count, level1::KLineType kline_type) {
            try {
                auto conn = level1::get_std_conn();
                //constexpr auto category = kline_type;
                level1::SecurityBarsRequest request(code, kline_type, start, count);
                level1::SecurityBarsResponse response(request.isIndex, kline_type);
                auto err = level1::process(conn->socket(), request, response);
                if (err) {
                    throw std::runtime_error(fmt::format("Process error: {}", err.message()));
                }
                return response.List;
            } catch (const std::exception &e) {  // 其他标准异常
                spdlog::error("[data::KLine] - 标准异常: {} (type: {})", e.what(), typeid(e).name());
                // 对于system_error可以记录更多信息
                if (auto se = dynamic_cast<const std::system_error *>(&e)) {
                    spdlog::error("[data::KLine] Error code: {}, category: {}", se->code().value(), se->code().category().name());
                }
            } catch (...) {
                spdlog::error("[data::KLine] 获取日K线异常");
            }
            return {};
        }
    } // namespace

    namespace {

        void save_kline_raw(const std::string &filename, const std::vector<KLineRaw>& values) {
            auto ec = filepath::check_filepath(filename, true);
            ec.clear();
            io::CSVWriter writer(filename);
            writer.write_row("date", "open", "close", "high", "low", "volume", "amount", "up", "down", "datetime");
            for (const auto &row: values) {
                writer.write_row(row.date, row.open, row.close, row.high, row.low, row.volume, row.amount,
                                 row.up, row.down, row.datetime);
            }
        }
    }

    // void KLineRaw::adjust(const factors::CumulativeAdjustment &adj) {
    //     auto m = adj.m;
    //     auto a = adj.a;
    //     this->Open = this->Open * m + a;
    //     this->Close = this->Close * m + a;
    //     this->High = this->High * m + a;
    //     this->Low = this->Low * m + a;
    //     // 成交量复权
    //     // 1. 计算均价
    //     auto ap = this->Amount / this->Volume;
    //     // 2. 均价复权
    //     ap = ap * m + a;
    //     // 3. 以成交金额为基准, 用复权均价计算成交量
    //     this->Volume = this->Amount / ap;
    // }

    std::vector<KLineRaw> read_kline_raw_from_csv(const std::string& filename) {
        std::vector<KLineRaw> klines;
        try {
            // 创建 CSV 读取器
            io::CSVReader<10> in(filename);  // 有 11 列数据

            // 设置表头字段名(用于自动匹配顺序)
            in.read_header(io::ignore_extra_column,
                           "date", "open", "close", "high", "low",
                           "volume", "amount", "up", "down", "datetime");

            KLineRaw row = {};
            while (in.read_row(row.date, row.open, row.close, row.high, row.low, row.volume, row.amount,
                               row.up, row.down, row.datetime)) {
                klines.emplace_back(row);
            }
        } catch(...) {
            // 忽略异常, 读csv文件失败, 返回空
        }
        return klines;
    }

    std::vector<KLineRaw> load_kline_raw(const std::string &code) {
        auto filename = config::get_kline_filename(code, false);
        spdlog::debug("[data::KLineRaw] kline file: {}", filename);
        return read_kline_raw_from_csv(filename);
    }

    void DataKLineRaw::Print(const std::string &code, const std::vector<exchange::timestamp> &dates)  {
        (void)code;
        (void)dates;
    }

    void DataKLineRaw::Update(const std::string &code, const exchange::timestamp &date) {
//        if(date != exchange::last_trading_day()) {
//            return;
//        }
        (void)date;
        // 1. 确定本地有效数据最后1条数据作为拉取数据的开始日期
        auto startDate = market_first_date;
        try {
            std::string cache_filename = config::get_kline_filename(code, false);
            std::vector<KLineRaw> cacheKLines = read_kline_raw_from_csv(cache_filename);
            auto kLength = cacheKLines.size();
            auto klineDaysOffset = detail::MAX_KLINE_LOOKBACK_DAYS;
            if(kLength > 0) {
                if (klineDaysOffset > kLength) {
                    klineDaysOffset = kLength;
                }
                startDate = cacheKLines[kLength-klineDaysOffset].date;
            }
            // 2. 确定结束日期
            auto endDate = exchange::timestamp::now().pre_market_time();
            spdlog::debug("[data::KLineRaw] [{}]: from {} to {}", code, startDate.only_date(), endDate.only_date());
            auto ts = exchange::date_range(startDate, endDate);
            auto total = u16(ts.size());
            startDate = ts[0];
            endDate = ts[total-1];
            spdlog::debug("[dataset::KLineRaw] [{}]: from {} to {}", code, startDate.only_date(), endDate.only_date());
            u16 step = level1::security_bars_max;
            u16 start = 0;
            //u16 category = level1::RI_K;
            // 3. 拉取数据
            std::vector<std::vector<level1::SecurityBar>> hs;
            //std::vector<level1::SecurityBar> history;
            size_t elementCount = 0;
            do {
                u16 count = step;
                if(total - start >= step) {
                    count = step;
                } else {
                    count = total - start;
                }
                auto reply = detail::fetch_kline(code, start, count);
                if (reply.empty()) {
                    break;
                }
                elementCount += reply.size();
                //hs.insert(hs.end(), reply.begin(), reply.end());
                hs.emplace_back(reply);
                if (reply.size() < count) {
                    break;
                }
                start += count;
            } while (start < total);
            // 4. 由于K线数据，每次获取数据是从后往前获取, 所以这里需要反转历史数据的切片
            std::reverse(hs.begin(), hs.end());
            // 5. 调整成交量, 单位从手改成股, vol字段 * 100
            std::vector<KLineRaw> incremental_klines;
            incremental_klines.reserve(elementCount);
            for(const auto & vec : hs) {
                for (const auto & row : vec) {
                    auto dateTime = exchange::timestamp(row.Year, row.Month, row.Day).pre_market_time();
                    if (dateTime < startDate || dateTime > endDate) {
                        continue;
                    }
                    auto kx = KLineRaw{
                        .date = dateTime.only_date(), // 日期
                        .open = row.Open,             // 开盘价
                        .close = row.Close,           // 收盘价
                        .high = row.High,             // 最高价
                        .low = row.Low,               // 最低价
                        .volume = row.Vol * 100,      // 成交量(股)
                        .amount = row.Amount,         // 成交金额(元)
                        .up = row.UpCount,            // 上涨家数 / 外盘
                        .down = row.DownCount,        // 下跌家数 / 内盘
                        .datetime = row.DateTime,     // 时间
                    };
                    incremental_klines.emplace_back(kx);
                }
            }
            // 6. K线数据转换成KLine结构
            // 6.1 判断是否已除权的依据是当前更新K线只有1条记录
//            bool adjusted = incremental_klines.size() == 1;
//            auto dividends = load_xdxr(code);
//            if (adjusted) {
//                calculate_pre_adjust(incremental_klines, startDate, dividends);
//            }
            // 6.2 只前复权当日数据
            // 7. 拼接缓存和新增的数据
            std::vector<KLineRaw> klines;
            // 7.1 先截取本地缓存的数据
            if (kLength > klineDaysOffset) {
                klines.insert(klines.end(), cacheKLines.begin(), cacheKLines.begin()+(kLength-klineDaysOffset));
            }
            // 7.2 拼接新增的数据
            if (klines.empty()) {
                klines = incremental_klines;
            } else {
                klines.insert(klines.end(), incremental_klines.begin(), incremental_klines.end());
            }
            // 8. 前复权
//            if(!adjusted) {
//                calculate_pre_adjust(klines, startDate, dividends);
//            }
            // 9. 刷新缓存文件
            save_kline_raw(cache_filename, klines);
        } catch (const std::exception &e) {  // 其他标准异常
            spdlog::error("[data::KLineRaw] - 标准异常: {} (type: {})", e.what(), typeid(e).name());
            // 对于system_error可以记录更多信息
            if (auto se = dynamic_cast<const std::system_error *>(&e)) {
                spdlog::error("[data::KLineRaw] Error code: {}, category: {}", se->code().value(), se->code().category().name());
            }
        } catch (...) {
            spdlog::error("[data::KLineRaw] 获取日K线异常");
        }
    }


} // namespace data

File: quant1x/data/kline_raw.py
# -*- coding: UTF-8 -*-
import os
import pandas as pd
from dataclasses import dataclass
from typing import List, Optional
import logging

from quant1x.exchange import Timestamp
from quant1x.level1 import protocol
from quant1x.level1.client import get_std_conn
from quant1x.level1.security_bars import SecurityBarsRequest, SecurityBarsResponse, KLineType, SecurityBar, SECURITY_BARS_MAX
from quant1x.config import config
from quant1x.data import adapter
from quant1x.data.adapter import DataAdapter, PLUGIN_MASK_BASE_DATA, register, DEFAULT_DATA_PROVIDER
from quant1x.data.base import BASE_RAW_DAILY_KLINE

logger = logging.getLogger(__name__)

MAX_KLINE_LOOKBACK_DAYS = 1

def validate_csv_columns(df: pd.DataFrame, required_columns: List[str],
                        filename: str = "", strict_order: bool = False) -> bool:
    """
    通用CSV列名验证函数

    Args:
        df: pandas DataFrame
        required_columns: 必需的列名列表
        filename: 文件名（用于日志）
        strict_order: 是否要求列的顺序严格匹配

    Returns:
        bool: 验证是否通过
    """
    # 清理列名（去除空格），但保持原始大小写
    actual_columns = clean_column_names(df)

    # 方法1: 检查必需列是否存在
    missing_cols = [col for col in required_columns if col not in actual_columns]
    if missing_cols:
        logger.error(f"CSV文件 {filename} 缺少必需列: {missing_cols}")
        logger.error(f"实际列名: {actual_columns}")
        logger.error(f"期望列名: {required_columns}")
        return False

    # 方法2: 检查列顺序（可选）
    if strict_order and actual_columns != required_columns:
        logger.warning(f"CSV文件 {filename} 列名顺序不匹配")
        logger.warning(f"期望顺序: {required_columns}")
        logger.warning(f"实际顺序: {actual_columns}")
        return False

    # 方法3: 检查额外列
    extra_cols = [col for col in actual_columns if col not in required_columns]
    if extra_cols:
        logger.info(f"CSV文件 {filename} 包含额外列: {extra_cols}")

    return True

def clean_column_names(df: pd.DataFrame) -> List[str]:
    """
    清理列名：只去除空格和不可见字符，保持原始大小写格式

    Args:
        df: 输入DataFrame

    Returns:
        清理后的列名列表
    """
    return df.columns.str.strip().tolist()

@dataclass
class KLineRaw:
    date: str = ""
    open: float = 0.0
    close: float = 0.0
    high: float = 0.0
    low: float = 0.0
    volume: float = 0.0
    amount: float = 0.0
    up: int = 0
    down: int = 0
    datetime: str = ""

    @staticmethod
    def headers() -> List[str]:
        return ["date", "open", "close", "high", "low", "volume", "amount", "up", "down", "datetime"]

    def adjust(self, factor):
        """
        根据复权因子调整K线价格数据
        
        Args:
            factor: CumulativeAdjustment对象，包含复权因子
        """
        if hasattr(factor, 'apply'):
            self.open = factor.apply(self.open)
            self.close = factor.apply(self.close)
            self.high = factor.apply(self.high)
            self.low = factor.apply(self.low)
        else:
            # 如果factor没有apply方法，假设它有m和a属性
            self.open = self.open * factor.m + factor.a
            self.close = self.close * factor.m + factor.a
            self.high = self.high * factor.m + factor.a
            self.low = self.low * factor.m + factor.a

def save_kline_raw(filename: str, values: List[KLineRaw]):
    if not values:
        return
        
    dirname = os.path.dirname(filename)
    if dirname:
        os.makedirs(dirname, exist_ok=True)
    
    data = [
        {
            "date": v.date,
            "open": v.open,
            "close": v.close,
            "high": v.high,
            "low": v.low,
            "volume": v.volume,
            "amount": v.amount,
            "up": v.up,
            "down": v.down,
            "datetime": v.datetime
        }
        for v in values
    ]
    
    df = pd.DataFrame(data, columns=KLineRaw.headers())
    df.to_csv(filename, index=False)

def read_kline_raw_from_csv(filename: str) -> List[KLineRaw]:
    """
    从CSV文件读取K线原始数据，包含完整的列名验证
    """
    klines = []
    if not os.path.exists(filename):
        return klines

    try:
        df = pd.read_csv(filename)

        # 使用通用列名验证函数，默认保持原始格式
        required_cols = KLineRaw.headers()
        if not validate_csv_columns(df, required_cols, filename, strict_order=True):
            return klines

        # 可选：如果需要处理列名问题，可以启用清理
        # df = clean_column_names(df)

        # 数据类型验证和转换
        for _, row in df.iterrows():
            try:
                kline = KLineRaw(
                    date=str(row['date']),
                    open=float(row['open']),
                    close=float(row['close']),
                    high=float(row['high']),
                    low=float(row['low']),
                    volume=float(row['volume']),
                    amount=float(row['amount']),
                    up=int(row['up']),
                    down=int(row['down']),
                    datetime=str(row['datetime'])
                )
                klines.append(kline)
            except (ValueError, TypeError) as e:
                logger.warning(f"跳过无效数据行: {e}")

    except Exception as e:
        logger.error(f"读取K线原始数据CSV文件失败 {filename}: {e}")

    return klines

def load_kline_raw(code: str) -> List[KLineRaw]:
    """
    从缓存文件加载指定证券代码的K线原始数据
    
    Args:
        code (str): 证券代码
        
    Returns:
        List[KLineRaw]: K线原始数据列表
    """
    cache_filename = config.get_kline_filename(code, forward=False)
    return read_kline_raw_from_csv(cache_filename)

def ensure_kline_raw_updated(code: str):
    """
    确保指定证券代码的K线原始数据是最新的
    
    Args:
        code (str): 证券代码
    """
    # 使用DataKLineRaw的update方法来确保数据是最新的
    data_adapter = DataKLineRaw()
    data_adapter.update(code)

def checkout_kline_raw(code: str) -> List[KLineRaw]:
    """
    获取指定证券的未复权K线数据，如果数据不存在则下载
    
    Args:
        code (str): 证券代码
        
    Returns:
        List[KLineRaw]: 未复权K线数据列表
    """
    # 确保数据是最新的
    ensure_kline_raw_updated(code)
    # 从缓存加载数据
    return load_kline_raw(code)

def fetch_kline(code: str, start: int, count: int, kline_type: KLineType = KLineType.DAILY) -> List[SecurityBar]:
    try:
        with get_std_conn() as conn:
            req = SecurityBarsRequest(code, kline_type.value, start, count)
            resp = SecurityBarsResponse(req.is_index, kline_type.value)
            protocol.process(conn, req, resp)
            return resp.list
    except Exception as e:
        logger.error(f"[dataset::KLineRaw] fetch_kline error: {e}")
        return []

class DataKLineRaw(DataAdapter):
    def kind(self) -> int:
        return BASE_RAW_DAILY_KLINE  # 基础数据-未复权K线

    def owner(self) -> str:
        return DEFAULT_DATA_PROVIDER

    def key(self) -> str:
        return "day_raw"

    def name(self) -> str:
        return "日K线RAW"

    def usage(self) -> str:
        return "日K线RAW数据适配器"

    def print(self, code: str, dates: Optional[List[Timestamp]] = None) -> None:
        """控制台打印K线数据"""
        klines = checkout_kline_raw(code)
        if not klines:
            print(f"No kline data found for {code}")
            return
            
        print(f"K线数据 for {code}:")
        for kline in klines[-10:]:  # 显示最近10条
            print(f"  {kline.date}: O:{kline.open:.2f} H:{kline.high:.2f} L:{kline.low:.2f} C:{kline.close:.2f} V:{kline.volume:.0f}")

    def update(self, code: str, date: Optional[Timestamp] = None) -> None:
        # 1. Determine start date from local cache
        current_start_date = Timestamp.parse("1990-12-19")  # market_first_date
        cache_filename = config.get_kline_filename(code, forward=False)
        cache_klines = read_kline_raw_from_csv(cache_filename)
        
        klines_length = len(cache_klines)
        klines_offset_days = MAX_KLINE_LOOKBACK_DAYS
        
        if klines_length > 0:
            if klines_offset_days > klines_length:
                klines_offset_days = klines_length
            
            kline = cache_klines[klines_length - klines_offset_days]
            current_start_date = Timestamp.parse(kline.date)
        
        # 2. Determine end date
        current_end_date = Timestamp.now().get_pre_market_time()
        logger.debug(f"[dataset::KLineRaw] [{code}]: from {current_start_date.only_date()} to {current_end_date.only_date()}")
        
        step = SECURITY_BARS_MAX
        start = 0
        hs: List[List[SecurityBar]] = []
        element_count = 0
        
        while True:
            count = step
            reply = fetch_kline(code, start, count)
            if not reply:
                break
                
            element_count += len(reply)
            hs.append(reply)
            
            last_bar = reply[-1]
            last_bar_date = Timestamp.parse(f"{last_bar.Year}-{last_bar.Month:02d}-{last_bar.Day:02d}").get_pre_market_time()
            
            if last_bar_date < current_start_date:
                break
                
            if len(reply) < count:
                break
                
            start += count
            
        hs.reverse()
        
        incremental_klines: List[KLineRaw] = []
        
        for vec in hs:
            for row in vec:
                date_time = Timestamp.parse(f"{row.Year}-{row.Month:02d}-{row.Day:02d}").get_pre_market_time()
                
                if date_time < current_start_date or date_time > current_end_date:
                    continue
                    
                kx = KLineRaw(
                    date=date_time.only_date(),
                    open=row.Open,
                    close=row.Close,
                    high=row.High,
                    low=row.Low,
                    volume=row.Vol * 100,  # Convert to shares
                    amount=row.Amount,
                    up=row.UpCount,
                    down=row.DownCount,
                    datetime=row.DateTime
                )
                incremental_klines.append(kx)
                
        # 7. Merge
        klines = []
        if klines_length > klines_offset_days:
            klines.extend(cache_klines[:klines_length - klines_offset_days])
            
        klines.extend(incremental_klines)
        
        # 9. Save
        save_kline_raw(cache_filename, klines)


# 自动注册DataKLineRaw插件
from data.adapter import register

# 创建并注册DataKLineRaw插件
_data_kline_raw_plugin = adapter.register(DataKLineRaw)

if __name__ == "__main__":
    # Example usage
    code = "sh600000"
    klines = checkout_kline_raw(code)
    print(f"Loaded {len(klines)} kline records for {code}")
    if klines:
        # 显示最近5条记录
        print("Recent 5 records:")
        for kline in klines[-5:]:
            print(f"  {kline.date}: O:{kline.open:.2f} H:{kline.high:.2f} L:{kline.low:.2f} C:{kline.close:.2f} V:{kline.volume:.0f}")